GH-167: Add to home screen banner in order confirmation

marlass opened the pull request:
Create service for add to home screen feature. New banner with prompt in the order confirmation.
Closes #167

WeizhengSap commented on projects/storefrontlib/src/lib/pwa/add-to-home-screen-btn/add-to-home-screen-btn.component.spec.ts:
```diff
@@ -2,15 +2,19 @@ import { async, ComponentFixture, TestBed } from '@angular/core/testing';
 
```
I made some changes on this test, and pushed my changes.

marlass commented on projects/storefrontlib/src/lib/pwa/add-to-home-screen-btn/add-to-home-screen-btn.component.html:
```diff
@@ -1,3 +1,3 @@
 <span (click)="prompt()">
-  <ng-content *ngIf="isEnabled && canPrompt"></ng-content>
+  <ng-content *ngIf="canPrompt$ | async"></ng-content>
 </span>
```
No. We wanted to prompt on click for any children. That's why we wrapped `ng-content`

WeizhengSap commented on projects/storefrontlib/src/lib/pwa/pwa.module.ts:
```diff
@@ -28,7 +34,14 @@ export function pwaConfigurationFactory(
       useFactory: pwaConfigurationFactory,
       deps: [Config]
-    }
+    },
```
why not just call "init" in the service's constructor. So, this part can be removed.

marlass replied:
Removed `init` call after service refactor.

WeizhengSap commented on projects/storefrontlib/src/lib/pwa/services/add-to-home-screen.service.ts:
```diff
@@ -0,0 +1,64 @@
+import { Injectable } from '@angular/core';
```
```
import { Injectable } from '@angular/core';
import { PWAModuleConfig } from '../pwa.module-config';
import { GlobalMessageType } from './../../global-message/models/message.model';
import { GlobalMessageService } from '../../global-message/facade/global-message.service';

@Injectable()
export class AddToHomeScreenService {
  private deferredEvent: any;
  canPrompt = false;

  constructor(
    private config: PWAModuleConfig,
    private globalMessageService: GlobalMessageService
  ) {
    if (this.config.pwa.addToHomeScreen) {
      this.init();
    }
  }

  init() {
    window.addEventListener('beforeinstallprompt', event => {
      event.preventDefault();
      this.deferredEvent = event;
      this.canPrompt = true;
    });

    window.addEventListener('appinstalled', () => {
      this.globalMessageService.add({
        type: GlobalMessageType.MSG_TYPE_CONFIRMATION,
        text: 'SAP Storefront was added to your home screen'
      });

      this.canPrompt = false;
      this.deferredEvent = null;
    });
  }

  firePrompt() {
    if (this.deferredEvent) {
      this.deferredEvent.prompt();
    }
  }
}
```
I rewrite this service. I feel it is more readable.

marlass replied:
Looks better. Changed to this version, but kept canPrompt as observable.

marlass commented:
@WeizhengSap 
1. `AddToHomeScreenBannerComponent` is moved to `PwaModule`.
2. Extracted common part from `AddToHomeScreenBtnComponent` and `AddToHomeScreenBannerComponent` to separate file. These components now extends it. Also moved tests from `AddToHomeScreenBtnComponent` to this new one. Removed the tests for btn component because currently component don't have logic (only extends the generic one).
3. Applied refactor suggestion for service.

Could you check it again?

WeizhengSap commented on projects/storefrontlib/src/lib/pwa/components/add-to-home-screen.component.ts:
```diff
@@ -0,0 +1,16 @@
+import { AddToHomeScreenService } from './../services/add-to-home-screen.service';
+import { Observable } from 'rxjs';
+
+export class AddToHomeScreenComponent implements OnInit {
```
can we make this class as 'abstract'?

marlass replied:
Good idea. Changed it.

WeizhengSap commented on projects/storefrontlib/src/lib/pwa/pwa.module.ts:
```diff
@@ -8,14 +8,21 @@ import {
   return { enabled: (pwaConfig.production && pwaConfig.pwa.enabled) || false };
 }
 
+export function pwaFactory(addToHomeScreenService) {
```
Can we remove this pwaFactory and this 
```
{
      provide: APP_INITIALIZER,
      useFactory: pwaFactory,
      deps: [AddToHomeScreenService],
      multi: true
    }
```
Just provide AddToHomeScreenService should be ok, right?

marlass replied:
It won't work. This service needs to be initialized on app initialization and should be created even when don't have any consumers. `beforeinstallprompt` event can be called in 1 second or after 1 minute. As far as I know this is the only solution.

WeizhengSap commented on projects/storefrontlib/src/lib/pwa/services/add-to-home-screen.service.ts:
```diff
@@ -0,0 +1,55 @@
+
+  private canPrompt = new BehaviorSubject<boolean>(false);
+
+  canPrompt$ = this.canPrompt.asObservable();
```
why we want an observable canPrompt$? A boolean variable is enough, right?

marlass replied:
Boolean isn't enough. Each consumer should react instantly when `beforeinstallprompt` event is fired. Button in a mobile menu is initialized on app start. The event can be triggered 30 second later and this component should be shown there right after.
